src/lib.rs

```diff
@@ -4,135 +4,159 @@ use core::sync::atomic;
 use core::marker::PhantomData;
 
 use std::boxed::Box;
-use std::mem;
 use std::sync::atomic::AtomicUsize;
 
 use std::sync::atomic::Ordering;
 
-pub struct Mlsp<T: ?Sized> {
-    local_count: *mut usize,
-    atomic_count: *mut AtomicUsize,
-    
-    ptr: *mut T,
-
-    phantom: PhantomData<T>
+struct Packet<T> {
+    atomic_count: AtomicUsize,
+    value: T
 }
 
-pub struct MlspPackage<T: ?Sized> {
-    atomic_count: *mut AtomicUsize,
-    
-    ptr: *mut T,
-
-    phantom: PhantomData<T>
+pub struct Mlsp<T> {
+    phantom: PhantomData<T>,
+    local_count: *mut Cell<usize>,
+    packet_ptr: *mut Packet<T>
 }
 
-impl<T> Mlsp<T> {
-    pub fn new(value: T) -> Self {
-        let local_count_cell = Cell::new(0);
-        let atomic_count_cell = Cell::new(AtomicUsize::new(0));
+pub struct MlspPackage<T> {
```

Comment:
note that this won't implement `Send` or `Sync`, unless you implement them manually.

Reply:
I had meant to do that but it totally slipped my mind